EIPS/eip-2348.md

```diff
@@ -0,0 +1,228 @@
+
+### Version Header
+
+For contracts with the first byte from `0x01` to `0xff`, or whose total length is less than 4 bytes,
+the contract is treated exactly as through it had been deployed to an account with version `0`. For
+these contracts none of the other subsections in this EIP apply.
+
+When deploying a contract if a contract starts with `0x00` and has a length 4 or later the first
+four bytes form a version header. If a version header is not recognized by the EVM the contract
+deployment transaction fails with out-of-gas.
+
+For this EIP, only header version `1` (contracts starting with the byte stream `0x00` `0x00` `0x00`
+`0x01`) is defined. Future EIPs may expand on the valid set of headers. This version indicates that
+next three validations are applied to the content of the contract, keeping all other semantics of
+the current "version 0" EVM contracts, including the same gas schedule.
+
+**unresolved** - How do we deal with executing with the header?
+
+- Should contract execution start at index 4 as PC=0,
+  - This causes EXTCODECOPY indexes to not match up
+- should contract execution start at index 4 as PC=4,
+  - This would require some possibly non-trivial EVM changes
+- should the version header be a multi-byte instruction which is a no-op? Contract starts at 0 and
+  PC=0
+  - This introduces a new opcode, may be the simplest.
+
+### `BEGINDATA`
+
+As described in [EIP-2327] a new opcode `BEGINDATA` (`0xb6`) is added that indicates the remainder
+of the contract should not be considered executable code.
+
+### Invalid Opcode Validation
+
+All data between the Version Header and either the `BEGINDATA` marker or the end of the contract if
+`BEGINDATA` is not present must represent a valid EVM program at all points of the data. Invalid
+opcode validation consists of the following process:
+
+- Iterate over the code bytes one by one.
+  - If the code byte is a multi-byte operation, skip the appropriate number of bytes.
+  - If the code byte is a valid opcode or designated invalid instruction (`0xfe`), continue.
+  - If the code byte is the `BEGINDATA` operation (`0xb6`) stop iterating.
+  - Otherwise, throw out-of-gas.
+
+As of the Istanbul upgrade all of the multi-byte operations are the `PUSHn` series of operations
+from `0x60` to `0x7f`. Future upgrades may add more multi-byte operations.
+
+As of the Istanbul upgrade the invalid opcodes are `0x0c` to `0x0f`, `0x1e`, `0x1f`, `0x21` to
+`0x2f`, `0x46` to `0x4f`, `0x5c` to `0x5f`, `0xa5` to `0xaf`, `0xb3` to `0xef`, `0xf6` to `0xf9`,
+`0xfb`, `0xfc`, and `0xfe`. Future upgrades will remove items from this list. Note that `0xb6` is
+referenced in this spec as the `BEGINDATA` marker, but is not part of any deployed upgrade. Also
+note that `0xfe` would remain as a reserved 'invalid instruction' that will still be permitted.
+
+### Static Jump Validations
+
+For every jump operation preceded by a `PUSHn` instruction the value of the data pushed on to the
```

Comment:
What about jump operations _not_ preceded by a `PUSHn` instruction?

Reply:
> What about jump operations _not_ preceded by a `PUSHn` instruction?

In that case we have to use data flow analysis to determine if the argument to `JUMP` is a constant specified by a `PUSHn`. I assume validation is a one time thing (before deploying the contract?) so building a data flow graph does not seem to be too expensive.

BTW, compilers only generate `PUSH2` in this case because ... there is a size limit to the smart contract.

Reply:
I would say we should not use this relatively weak heuristics as a part of new validation rules. It is better to implement subroutines (which eliminates the most common source of dynamic jumps - which is return from the subroutine), and then the actual static jumps, and disable the dynamic jumps all together. Then we can remove `JUMPDEST`.

Reply:
Using subroutines is way better for validating the contract, but it is not infeasible to validate a contract without static jumps. Symbolic execution is still able to figure out which jump is dynamic and hence report it.